### PR TITLE
[5.4] Allow easy overriding of ServiceProvider configuration merging

### PR DESCRIPTION
**Pull request by willrowe**
I put this together after seeing #16991. Requests for recursive merging have come up several times and I thought this could be a good solution to allow for more advanced merging when needed.

If `mergeConfig` is not overridden in the `ServiceProvider` it will function exactly as it did before by doing a top-level (non-recursive) merge.

I have had several projects in the past where I needed to merge settings. In most cases a simple `array_merge` was sufficient, but sometimes I would need to have more control over nested values. Instead of using a catch-all solution that would automatically merge items recursively, I found it much easier to override the rare occurrences and handle them individually.

### Example based on #16991:
*package/src/ServiceProvider.php*
```php
<?php

namespace App;

use Illuminate\Support\ServiceProvider;

class TestServiceProvider extends ServiceProvider
{
    /**
     * @inheritDoc
     */
    protected function mergeConfig($key, array $baseConfig, array $mergeConfig)
    {
        // merge all the values like normal
        $config = parent::mergeConfig($key, $baseConfig, $mergeConfig);
        
        // merge the individual nested values as needed
        $config['other'] = array_merge($baseConfig['other'], $mergeConfig['other']);
        
        return $config;
    }
}
```
*config/package.php*
```php
<?php

return [
    'example1' => true,
    'other' => [
         'config1' => true,
         'config2' => true,
    ]
];
```
*package/config/main.php*
```php
<?php

return [
    'example1' => false,
    'example2' => false,
    'other' => [
         'config1' => false,
         'config2' => false,
         'config3' => false,
    ]
];
```
*test it out*
```php
$this->mergeConfigFrom('package/config/main.php', 'package');
dd(config('package'));
```

*result*
```
array:3 [▼
  "example1" => true
  "example2" => false
  "other" => array:3 [▼
    "config1" => true
    "config2" => true
    "config3" => false
  ]
]
```

### CODE REVIEW

**Comment by GrahamCampbell** [comment]
👎 from me. Just as easy to just override the whole method tbh.

**Comment by willrowe** [comment]
I thought the same thing too until I took a closer look at the `mergeConfigFrom` method and saw that the code to load the configurations would need to be duplicated. This way the method that needs to be overridden only handles the merging and not loading anything.

File: src/Illuminate/Support/ServiceProvider.php [hunk 1]
```diff
@@ -46,17 +46,34 @@ public function __construct($app)
...
 
-        $this->app['config']->set($key, array_merge(require $path, $config));
+    /**
+     * Merge two configurations together.
+     * @param  string $key
+     * @param  array  $baseConfig
+     * @param  array  $mergeConfig
+     * @return array
+     */
+    protected function mergeConfig($key, array $baseConfig, array $mergeConfig)
```
Comment (crynobone) [inline src/Illuminate/Support/ServiceProvider.php]: Why do we need `$key` here?

Comment (willrowe) [inline src/Illuminate/Support/ServiceProvider.php]: If there are multiple keys in a service provider you might need to handle each one differently. You'd be able to identify which one was being merged based on `$key`.

**Comment by GrahamCampbell** [comment]
> the code to load the configurations would need to be duplicated

Meh, it's only one line.

**Pull request closed without merging**